# November LeetCoding Challenge
### Smallest Integer Divisible by K
Given a positive integer K, you need to find the length of the smallest positive integer N such that N is divisible by K, and N only contains the digit 1.

Return the length of N. If there is no such N, return -1.

Note: N may not fit in a 64-bit signed integer.

 

Example 1:

Input: K = 1
Output: 1
Explanation: The smallest answer is N = 1, which has length 1.
Example 2:

Input: K = 2
Output: -1
Explanation: There is no such positive integer N divisible by 2.
Example 3:

Input: K = 3
Output: 3
Explanation: The smallest answer is N = 111, which has length 3.
 

Constraints:

1 <= K <= 105

In [1]:
for i in range (1, 10):
    string_val = "".join('1' for i in range(i))
    print(string_val)

1
11
111
1111
11111
111111
1111111
11111111
111111111


In [2]:
def smallestRepunitDivByK(K: int) -> int:
    answer = []
    for i in range (1, K+1):
        string_val = "".join('1' for i in range(i))

        if int(string_val) % K == 0:
            answer.append(len(string_val))

    if len(answer) > 0:
        return answer[0]
    else:
        return -1
    

In [3]:
for i in range (1, 20):
    print(str(i)+', ' +str(smallestRepunitDivByK(i)))

1, 1
2, -1
3, 3
4, -1
5, -1
6, -1
7, 6
8, -1
9, 9
10, -1
11, 2
12, -1
13, 6
14, -1
15, -1
16, -1
17, 16
18, -1
19, 18


#### Too Slow 50/70

In [4]:
for i in range (1, 10):
    string_val = '1'*i
    print(string_val)

1
11
111
1111
11111
111111
1111111
11111111
111111111


In [5]:
def smallestRepunitDivByK(K: int) -> int:
    answer = []
    for i in range (1, K+1):
        string_val = '1'*i

        if int(string_val) % K == 0:
            answer.append(len(string_val))

    if len(answer) > 0:
        return answer[0]
    else:
        return -1
    

In [6]:
for i in range (1, 20):
    print(str(i)+', ' +str(smallestRepunitDivByK(i)))

1, 1
2, -1
3, 3
4, -1
5, -1
6, -1
7, 6
8, -1
9, 9
10, -1
11, 2
12, -1
13, 6
14, -1
15, -1
16, -1
17, 16
18, -1
19, 18


#### Still too Slow 60/70

In [129]:
def smallestRepunitDivByK(K: int) -> int:
    answer = []
    K2 = 1
    
    while  K>= K2 and len(answer) == 0:
        
        string_val = '1'*K2
        if int(string_val) % K == 0:
            answer.append(len(string_val))
            break
        K2 += 1

    if len(answer) > 0:
            return answer[0]
    else:
        return -1

     
    

In [12]:
smallestRepunitDivByK(7)

6

In [13]:
for i in range (1, 20):
    print(str(i)+', ' +str(smallestRepunitDivByK(i)))

1, 1
2, -1
3, 3
4, -1
5, -1
6, -1
7, 6
8, -1
9, 9
10, -1
11, 2
12, -1
13, 6
14, -1
15, -1
16, -1
17, 16
18, -1
19, 18


#### Still too Slow; 61 / 70 test cases passed.

In [31]:
def smallestRepunitDivByK(K: int) -> int:
    answer = []
    K2 = 1
    
    while  K>= K2 and len(answer) == 0:
        
        string_val = '1'*K2
        if int(string_val) % K == 0:
            answer.append(len(string_val))
            break
        K2 += 1

    if len(answer) > 0:
            return answer[0]
    else:
        return -1

In [126]:
for i in range (1, 106):
    print(str(i)+', ' +str(smallestRepunitDivByK(i)))

1, 1
2, -1
3, 3
4, -1
5, -1
6, -1
7, 6
8, -1
9, 9
10, -1
11, 2
12, -1
13, 6
14, -1
15, -1
16, -1
17, 16
18, -1
19, 18
20, -1
21, 6
22, -1
23, 22
24, -1
25, -1
26, -1
27, 27
28, -1
29, 28
30, -1
31, 15
32, -1
33, 6
34, -1
35, -1
36, -1
37, 3
38, -1
39, 6
40, -1
41, 5
42, -1
43, 21
44, -1
45, -1
46, -1
47, 46
48, -1
49, 42
50, -1
51, 48
52, -1
53, 13
54, -1
55, -1
56, -1
57, 18
58, -1
59, 58
60, -1
61, 60
62, -1
63, 18
64, -1
65, -1
66, -1
67, 33
68, -1
69, 66
70, -1
71, 35
72, -1
73, 8
74, -1
75, -1
76, -1
77, 6
78, -1
79, 13
80, -1
81, 81
82, -1
83, 41
84, -1
85, -1
86, -1
87, 84
88, -1
89, 44
90, -1
91, 6
92, -1
93, 15
94, -1
95, -1
96, -1
97, 96
98, -1
99, 18
100, -1
101, 4
102, -1
103, 34
104, -1
105, -1


In [41]:
for i in range(1, 8):
    print(int('1'*i)%7)


1
4
6
5
2
0
1


11111 = 1111 * 10 + 1 We only need to store remainders modulo K.
If we never get a remainder of 0, why would that happen, and how would we know that?

In [45]:
K = 10

In [46]:
lisst = []
for i in range(1,K):
    lisst.append(int('1'*i))
    

for i in range(1,10):
    for j in range(1,10):
        print(str(i)+"x"+str(j))
        print(j*i)

for K in range(1,11):
    print(K%10)

In [109]:
for i in range(1,105):
    for j in range(1,105):
        if (j*i)%1000 == 111:
            print(str(i)+"x"+str(j))
            print(j*i)

SyntaxError: invalid syntax (<ipython-input-109-e12ccd83fde0>, line 2)

In [125]:
def smallestRepunitDivByK(K: int) -> int:
    
    if K not in [1,3,7,9,11,13,17,19,21,23,27,29,31,33,37,39,41,43,47,49,51,53,57,59,61,63,67,69,71,73,77,79,81,83,87,89,91,93,97,99,101,103]:
        return -1
    
    else:
        for i in range (1, K+1):
            string_val = '1'*i

            if int(string_val) % K == 0:
                
                break
        return len(string_val)

In [131]:
answer= []
for i in range (1, 10**5):
    if smallestRepunitDivByK(i) != -1:
        answer.append(i)

KeyboardInterrupt: 

In [ ]:
answer

In [ ]:
class Solution:
    def uniqueMorseRepresentations(self, words: List[str]) -> int:

        morse =    [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
        letter_mapping = dict(zip(ascii_lowercase, morse))
        answer = []
        for word in words:
            for letter in word:
                letter.apply.letter_mapping

In [171]:
def letter_indexes(text):
    letter_mapping = dict(zip(ascii_lowercase, morse))
    indexes = [
      letter_mapping[letter] for letter in text.lower() 
      if letter in letter_mapping
    ]

    return ''.join(str(index) for index in indexes)

In [145]:
from string import ascii_lowercase
morse =    [".-","-...","-.-.","-..",".","..-.","--.","....","..",".---","-.-",".-..","--","-.","---",".--.","--.-",".-.","...","-","..-","...-",".--","-..-","-.--","--.."]
answer = []
--letter_indexes(["gin", "zen", "gig", "msg"])

AttributeError: 'list' object has no attribute 'lower'

In [175]:
answer = []
for word in ["gin", "zen", "gig", "msg"]:
    if letter_indexes(word) not in answer:
        answer.append(letter_indexes(word))

print(len(answer))

2


In [180]:
def letter_indexes(text):
    letter_mapping = dict(zip(ascii_lowercase, morse))
    indexes = [
      letter_mapping[letter] for word in text for letter in word 
      if letter in letter_mapping
    ]

    return ''.join(str(index) for index in indexes)

In [181]:
letter_indexes(["gin", "zen", "gig", "msg"])

'--...-.--...-.--...--.--...--.'